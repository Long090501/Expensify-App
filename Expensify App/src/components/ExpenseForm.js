import React from 'react'
import moment from 'moment'
import { SingleDatePicker } from 'react-dates'

class ExpenseForm extends React.Component {
    constructor(props) {
        super(props)
        this.state = {
            description: props.expense ? props.expense[0].description : '',
            amount: props.expense ? (props.expense[0].amount).toString() : '',
            note: props.expense ? props.expense[0].note : '',
            createdAt: props.expense ? moment(props.expense[0].createdAt) : moment(),
            calendarFocused: false,
            error: ''
        }
    }

    descriptionChange = (e) => {
        const description = e.target.value
        this.setState(() => ({description}))
    } 

    amountChange = (e) => {
        const amount = e.target.value
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({amount}))
        }
    } 

    noteChange = (e) => {
        const note = e.target.value
        this.setState(() => ({note}))
    }

    onDateChange = (createdAt) => {
        if(createdAt) {
            this.setState(() => ({createdAt}))
        }
    }

    onFocusChange = ({focused}) => {
        this.setState(() => ({calendarFocused: focused}))
    }

    onSubmit = (e) => {
        e.preventDefault()
        if(!this.state.description || !this.state.amount) {
            this.setState(() => ({error: 'Please provide description and amount info!'}))
        }
        else {
            this.setState(() => ({error: ''}))
            this.props.handleSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10),
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    }

    render() {
        return (
        <div>
            {this.state.error && <h3>{this.state.error}</h3>}
            <form onSubmit = {this.onSubmit}>
                <input type = 'text' placeholder = 'Description' value = {this.state.description} autoFocus
                onChange = {this.descriptionChange}
                ></input>
                <input type = 'text' placeholder = 'Amount' value = {this.state.amount} autoFocus
                onChange = {this.amountChange}></input>
                <textarea type = 'text' placeholder = 'Add a note for your expense (optional)' value = {this.state.note} autoFocus
                onChange = {this.noteChange}></textarea>
                <SingleDatePicker
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
                <button>Add Expense</button>
            </form>
        </div>
        )
    }
}

export {ExpenseForm}