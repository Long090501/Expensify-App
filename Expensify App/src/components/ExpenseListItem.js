import React from 'react'
import { connect } from 'react-redux'
import { removeExpense } from '../actions/expenses'
import { Link } from 'react-router-dom'

    const ExpenseListItem = (props) => {
    return (
    <div>
    <Link to = {`/edit/${props.expense.id}`}> 
        <h3>{props.expense.description}</h3>
    </Link>
        <p>{props.expense.amount} - {props.expense.createdAt}</p>
        <button onClick = {() => {
            props.dispatch(removeExpense(props.expense.id))
        }}>Remove</button>
    </div>
    )
}

const ConnectedExpenseListItem = connect()(ExpenseListItem)
export {ConnectedExpenseListItem}