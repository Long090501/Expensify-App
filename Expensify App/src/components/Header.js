import React from 'react'
import { NavLink } from 'react-router-dom'

const Header = () => (
    <div>
    <h1>Expensify App</h1>
    <NavLink end to = '/' className={({ isActive }) =>
              isActive ? 'is-active' : undefined
            }>Dashboard</NavLink>
    <NavLink to = '/create' className={({ isActive }) =>
              isActive ? 'is-active' : undefined
            }>Create Expense</NavLink>
    <NavLink to = '/help' className={({ isActive }) =>
              isActive ? 'is-active' : undefined
            }>Help</NavLink>
    </div>
)

export {Header}