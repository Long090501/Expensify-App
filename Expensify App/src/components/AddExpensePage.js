import React from 'react'
import {ExpenseForm} from './ExpenseForm'
import { connect } from 'react-redux'
import { addExpense } from '../actions/expenses'

    const AddExpensePage = (props) => (
    <div>
        <h1>Add Expense</h1>
        <ExpenseForm handleSubmit = {(expense) => {
            props.onSubmit(expense)
        }}/>
    </div>
)

const mapDispatchToProps = (dispatch) => ({
    onSubmit: (expense) => dispatch(addExpense(expense))
})

const ConnectedAddExpensePage = connect(undefined, mapDispatchToProps)(AddExpensePage)
export { ConnectedAddExpensePage }