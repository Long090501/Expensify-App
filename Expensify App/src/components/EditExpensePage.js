import React from 'react'
import { useSelector, useDispatch } from 'react-redux'
import { useParams } from 'react-router-dom'
import {ExpenseForm} from './ExpenseForm'
import { editExpense } from '../actions/expenses'

const EditExpensePage = () => {
    const getID = useParams()
    const dispatch = useDispatch()
    const expenses = useSelector(state => state.expenses)
    const expense = expenses.find((element) => {
        if(element[0] && getID) {
            return element[0].id === getID.id
        }
    })

    return ( 
    <div>
        <ExpenseForm expense = {expense} handleSubmit = {(element) => {    
            dispatch(editExpense(expense[0].id, element))
        }}></ExpenseForm>
    </div>
    )
}

export {EditExpensePage}