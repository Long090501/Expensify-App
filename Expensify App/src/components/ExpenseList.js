import React from 'react'
import { connect } from 'react-redux'
import {ConnectedExpenseListItem} from './ExpenseListItem'
import {ConnectedExpenseListFilters} from './ExpenseListFilters'
import {filteringData} from '../selectors/filtering'

export const ExpenseList = (props) => (
    <div>
        <h1>Expense List</h1>
        <ConnectedExpenseListFilters></ConnectedExpenseListFilters>
        {props.expenses.map((expense, index) => {
            return <ConnectedExpenseListItem key = {index} expense = {expense[0]}></ConnectedExpenseListItem>
        })}
    </div>
)

const mapStateToProps = (state) => {
    return {
        expenses: filteringData(state.expenses, state.filters), 
        filters: state.filters,
    }
} 

const ConnectedExpenseList = connect(mapStateToProps)(ExpenseList)
export {ConnectedExpenseList}