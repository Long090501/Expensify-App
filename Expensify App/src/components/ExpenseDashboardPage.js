import React from 'react'
import {ConnectedExpenseList} from './ExpenseList'

const ExpenseDashboardPage = () => (
    <div>
        <ConnectedExpenseList></ConnectedExpenseList>
    </div>
)

export { ExpenseDashboardPage }