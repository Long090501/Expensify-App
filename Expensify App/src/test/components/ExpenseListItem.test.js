import React from 'react'
import { shallow } from 'enzyme'
import moment from 'moment'
import { ExpenseListItem } from '../../components/ExpenseListItem'

const expenses = [
    [{
        id: 1,
        description: 'gas bill',
        note: 'a',
        amount: 500,
        createdAt: moment(0).subtract(3, 'days').valueOf()
    }],
    [{
        id: 2,
        description: 'water bill',
        note: 'b',
        amount: 300,
        createdAt: 0
    }],
    [{
        id: 3,
        description: 'internet bill',
        note: 'c',
        amount: 900,
        createdAt: moment(0).add(3, 'days').valueOf()
    }]
]

test('should render expense list item correctly', () => {
    const wrapper = shallow(<ExpenseListItem expense = {expenses[1][0]}/>)
    expect(wrapper).toMatchSnapshot()
})