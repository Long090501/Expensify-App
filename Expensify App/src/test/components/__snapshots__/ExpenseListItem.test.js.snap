// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense list item correctly 1`] = `
<div>
  <Link
    to="/edit/2"
  >
    <h3>
      water bill
    </h3>
  </Link>
  <p>
    300
     - 
    0
  </p>
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
